File: volume_1/chapter_7/langchain/llm_tools_gateway.py
"""
LLM Structured Gateway - LangChain with tool triggering via tools.py
"""

import sys
import os
import re
from langchain_anthropic import ChatAnthropic
from langchain_openai import ChatOpenAI
from langchain_google_genai import ChatGoogleGenerativeAI
from langchain_core.messages import HumanMessage, SystemMessage

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..', 'chapter_4')))

from utils import get_api_key, get_default_model, BaseLLMManager, interactive_cli
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
from tools import run_tool


class LangChainLLMManager(BaseLLMManager):
    def __init__(self):
        super().__init__("LangChain+Structured")

    def _create_client(self, provider: str, temperature: float, max_tokens: int):
        if provider == "anthropic":
            return ChatAnthropic(
                anthropic_api_key=get_api_key(provider),
                model=get_default_model(provider),
                temperature=temperature,
                max_tokens=max_tokens
            )
        elif provider == "openai":
            return ChatOpenAI(
                openai_api_key=get_api_key(provider),
                model=get_default_model(provider),
                temperature=temperature,
                max_tokens=max_tokens
            )
        elif provider == "google":
            return ChatGoogleGenerativeAI(
                google_api_key=get_api_key(provider),
                model=get_default_model(provider),
                temperature=temperature,
                max_output_tokens=max_tokens
            )
        elif provider == "xai":
            return ChatOpenAI(
                openai_api_key=get_api_key(provider),
                openai_api_base="https://api.x.ai/v1",
                model=get_default_model(provider),
                temperature=temperature,
                max_tokens=max_tokens
            )
        else:
            raise ValueError(f"Unsupported provider: {provider}")

    def _test_provider(self, provider: str):
        self._create_client(provider, temperature=0.7, max_tokens=1000)

    def _extract_tool_request(self, text: str):
        match = re.search(r'TOOL:\s*(\w+)\s+with\s+(.*)', text)
        if not match:
            return None
        name = match.group(1)
        params = match.group(2)
        args = {}
        for part in params.split(','):
            if '=' in part:
                key, val = part.split('=', 1)
                args[key.strip()] = val.strip().strip("\"'")
        return name, args

    def ask_question(self, topic: str, provider: str = None,
                     template: str = "{topic}", max_tokens: int = 1000,
                     temperature: float = 0.7, session_id: str = "default") -> dict:

        prompt = template.format(topic=topic)
        available_providers = self.get_available_providers()
        if not provider or provider not in available_providers:
            if not available_providers:
                return {
                    "success": False,
                    "error": "No providers available",
                    "provider": "none",
                    "model": "none",
                    "prompt": prompt,
                    "response": None
                }
            provider = available_providers[0]

        model = get_default_model(provider)

        try:
            client = self._create_client(provider, temperature, max_tokens)
            messages = [
                SystemMessage(content="You are a helpful assistant. If a tool is needed, say: TOOL: <name> with key=value"),
                HumanMessage(content=prompt)
            ]
            result = client.invoke(messages)
            text = result.content

            tool_call = self._extract_tool_request(text)
            if tool_call:
                tool_name, tool_args = tool_call
                tool_output = run_tool(tool_name, tool_args)
                text += f"\n\n[TOOL OUTPUT]: {tool_output}"

            return {
                "success": True,
                "provider": provider,
                "model": model,
                "prompt": prompt,
                "response": text,
                "temperature": temperature,
                "max_tokens": max_tokens,
                "session_id": session_id
            }

        except Exception as e:
            return {
                "success": False,
                "provider": provider,
                "model": model,
                "prompt": prompt,
                "error": str(e),
                "response": None,
                "temperature": temperature,
                "max_tokens": max_tokens,
                "session_id": session_id
            }


def main():
    if len(sys.argv) > 1 and sys.argv[1] == "web":
        from web import run_web_server
        run_web_server(lambda: LangChainLLMManager())
    else:
        manager = LangChainLLMManager()
        interactive_cli(manager)


if __name__ == "__main__":
    main()

File: volume_1/chapter_7/langchain/llm_tools_gateway.js
/**
 * LLM Tools Gateway - LangChain JS (Chapter 7).
 *
 * Chapter flow: 4 (providers) -> 5 (memory) -> 6 (structured) -> 7 (tools).
 */

import { LangChainLLMManager as Chapter6LangChainManager } from '../../chapter_6/langchain/llm_structured_gateway.js';
import { interactiveCli, getDefaultModel } from '../../chapter_4/utils.js';
import { buildToolsPrompt, runTool } from '../tools.js';

const TOOLS_TEMPLATE = `You are a helpful assistant with access to external tools.

Available tools:
{tools}

For every response, return strict JSON with this shape:
{
  "tool_call": null OR {"name": "tool_name", "arguments": {"arg": "value"}},
  "final_answer": "string"
}

Rules:
- If no tool is needed, set tool_call to null.
- If a tool is needed, set tool_call and keep final_answer short (what you expect to answer after tool execution).
- Return JSON only.

User topic: {topic}`;

const FOLLOW_UP_TEMPLATE = `You already requested a tool and now have the result.

Original user topic: {topic}
Tool call: {tool_call}
Tool output: {tool_output}

Return strict JSON:
{
  "tool_call": null,
  "final_answer": "final response for the user"
}`;

class LangChainLLMManager extends Chapter6LangChainManager {
    constructor(memoryEnabled = true) {
        super(memoryEnabled);
        this.framework = 'LangChain+Tools JS';
    }

    _extractJsonObject(raw) {
        let text = String(raw || '').trim();
        if (text.startsWith('```json')) text = text.slice(7);
        if (text.startsWith('```')) text = text.slice(3);
        if (text.endsWith('```')) text = text.slice(0, -3);
        text = text.trim();

        try {
            return JSON.parse(text);
        } catch {
            const match = text.match(/\{[\s\S]*\}/);
            if (!match) throw new Error('No JSON object found in model response');
            return JSON.parse(match[0]);
        }
    }

    async _invokeJsonStep(provider, prompt, temperature, maxTokens) {
        const client = this._createClient(provider, temperature, maxTokens);
        const result = await client.invoke(this._buildMessages(prompt));
        const text = this._extractText(provider, result);
        return this._extractJsonObject(text);
    }

    async askQuestion(topic, provider = null, template = TOOLS_TEMPLATE, maxTokens = 1000, temperature = 0.2, sessionId = 'default') {
        const prompt = template.replace('{topic}', topic).replace('{tools}', buildToolsPrompt());
        const resolvedProvider = this._resolveProvider(provider);

        if (!resolvedProvider) {
            return {
                success: false,
                error: 'No providers available',
                provider: 'none',
                model: 'none',
                prompt,
                response: null,
            };
        }

        const model = getDefaultModel(resolvedProvider);

        try {
            const firstStep = await this._invokeJsonStep(resolvedProvider, prompt, temperature, maxTokens);
            const toolCall = firstStep.tool_call;
            let finalAnswer = String(firstStep.final_answer || '').trim();
            let toolOutput = null;

            if (toolCall && typeof toolCall === 'object' && toolCall.name) {
                const toolName = String(toolCall.name);
                const toolArgs = toolCall.arguments && typeof toolCall.arguments === 'object' ? toolCall.arguments : {};
                toolOutput = runTool(toolName, toolArgs);

                const followUpPrompt = FOLLOW_UP_TEMPLATE
                    .replace('{topic}', topic)
                    .replace('{tool_call}', JSON.stringify(toolCall))
                    .replace('{tool_output}', String(toolOutput));

                const secondStep = await this._invokeJsonStep(resolvedProvider, followUpPrompt, temperature, maxTokens);
                finalAnswer = String(secondStep.final_answer || finalAnswer).trim() || finalAnswer;
            }

            return {
                success: true,
                provider: resolvedProvider,
                model,
                prompt,
                response: {
                    tool_call: toolCall ?? null,
                    tool_output: toolOutput,
                    final_answer: finalAnswer,
                },
                rawAnswer: finalAnswer,
                temperature,
                maxTokens,
                sessionId,
            };
        } catch (error) {
            return {
                success: false,
                provider: resolvedProvider,
                model,
                prompt,
                error: error.message,
                response: null,
                temperature,
                maxTokens,
                sessionId,
            };
        }
    }
}

async function main() {
    const args = process.argv.slice(2);
    if (args.length > 0 && args[0] === 'web') {
        const { runWebServer } = await import('../../chapter_4/web.js');
        await runWebServer(() => new LangChainLLMManager(true));
    } else {
        const manager = new LangChainLLMManager(true);
        await manager._checkProviders();
        await interactiveCli(manager);
    }
}

export { LangChainLLMManager };

if (import.meta.url === `file://${process.argv[1]}`) {
    main().catch(console.error);
}
